drivers/qmc5883l/qmc5883l.c

```diff
@@ -0,0 +1,203 @@
+        if (i2c_acquire(dev->i2c) != 0) {
+            return QMC5883L_BUSERR;
+        }
+    }
+    int res = i2c_write_reg(dev->i2c, ADDR, reg, val, 0);
+    if ((release) || (res != 0)) {
+        i2c_release(dev->i2c);
+    }
+    return (res == 0) ? QMC5883L_OK : QMC5883L_BUSERR;
+}
+
+int qmc5883l_init(qmc5883l_t *dev, const qmc5883l_params_t *params)
+{
+    assert(dev);
+    assert(params);
+    int res;
+
+    dev->i2c = params->i2c;
+    dev->pin_drdy = params->pin_drdy;
+    dev->cfg = (params->odr | params->rng | params->osr | QMC5883L_CONT);
+
+    /* lets start with a soft reset */
+    res = _reg_write(dev, QMC5883L_CTRL2, QMC5883L_SOFT_RST, 1, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+
+    /* verify the reset by reading the status register, should be all zeros */
+    uint8_t tmp;
+    if (i2c_read_reg(dev->i2c, ADDR, QMC5883L_STATUS, &tmp, 0) != 0) {
+        i2c_release(dev->i2c);
+        return QMC5883L_BUSERR;
+    }
+    if (tmp != 0) {
+        i2c_release(dev->i2c);
+        return QMC5883L_NOCFG;
+    }
+
+    /* write the actual device configuration */
+    res = _reg_write(dev, QMC5883L_SETRESET, 0x01, 0, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    res = _reg_write(dev, QMC5883L_CTRL2, QMC5883L_INT_ENB, 0, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    return _reg_write(dev, QMC5883L_CTRL1, (dev->cfg | QMC5883L_CONT), 0, 1);
+}
+
+int qmc5883l_data_ready(const qmc5883l_t *dev)
+{
+    assert(dev);
+    uint8_t status;
+
+    int res = _reg_read(dev, QMC5883L_STATUS, &status, 1, 1);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    return (status & QMC5883L_DRDY) ? QMC5883L_OK : QMC5883L_NODATA;
+}
+
+int qmc5883l_read(const qmc5883l_t *dev, int16_t *data_out)
+{
+    assert(data_out);
+    int16_t tmp[3];
+
+    int res = qmc5883l_read_raw(dev, tmp);
+    if ((res == QMC5883L_OK) || (res == QMC5883L_OVERFLOW)) {
+        unsigned scale = (dev->cfg & QMC5883L_RNG_8G) ? 8000 : 2000;
+        for (unsigned i = 0; i < 3; i++) {
+            data_out[i] = (int16_t)((((int32_t)tmp[i]) * scale) / INT16_MAX);
+        }
+    }
+
+    return res;
+}
+
+int qmc5883l_read_raw(const qmc5883l_t *dev, int16_t *data_out)
+{
+    assert(dev);
+    assert(data_out);
+
+    int res;
+    uint8_t status;
+    uint8_t tmp[6];
+
+    res = _reg_read(dev, QMC5883L_STATUS, &status, 1, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    if (!(status & QMC5883L_DRDY)) {
+        res = QMC5883L_NODATA;
+        goto done;
+    }
+    if (status & QMC5883L_OVL) {
+        res = QMC5883L_OVERFLOW;
+    }
+    if (i2c_read_regs(dev->i2c, ADDR, QMC5883L_DOXL, tmp, 6, 0) != 0) {
+        res = QMC5883L_BUSERR;
+        goto done;
+    }
+
+    /* convert data to host byte order */
+    data_out[0] = (int16_t)tmp[1] << 8 | tmp[0];
+    data_out[1] = (int16_t)tmp[3] << 8 | tmp[2];
+    data_out[2] = (int16_t)tmp[5] << 8 | tmp[4];
+
+done:
```

Comment:
Ups, the first lesson I learned in C programming 30 years ago was that C is not BASIC and goto is an absolute no go. I know, it is sometimes easier to work with labels and jumps, but ... It's probably a question of taste as with macros.

Reply:
That lesson is not 100% valid. On certain places the use of `goto` is ok, this being one of them. Have a look a Linux driver code, you'll find many gotos :-)

Reply:
That's Ok for me.